File: android/beacon/3-Alert/Complete/app/src/main/java/android/connecthings/com/tuto/alert/TimeConditions.java
package android.connecthings.com.tuto.alert;

import android.os.SystemClock;

import com.connecthings.connectplace.actions.inappaction.conditions.InAppActionConditionsDefault;
import com.connecthings.connectplace.actions.inappaction.conditions.parameter.InAppActionConditionsParameter;
import com.connecthings.util.Log;

/**
 * Created by Connecthings on 14/11/2017.
 */
public class TimeConditions extends InAppActionConditionsDefault {
    private long timeBeforeCreatingAlert;

    public TimeConditions(int maxTimeBeforeReset, long timeBeforeCreatingAlert) {
        super(maxTimeBeforeReset);
        this.timeBeforeCreatingAlert = timeBeforeCreatingAlert;
    }

    @Override
    public Class<? extends InAppActionConditionsParameter> getInAppActionParameterClass() {
        return TimeConditionsParameter.class;
    }

    @Override
    public String getConditionsKey() {
        return "inAppActionTimeConditions";
    }

    @Override
    public boolean isConditionValid(InAppActionConditionsParameter inAppActionConditionsParameter) {
        TimeConditionsParameter timeConditionsParameter = (TimeConditionsParameter) inAppActionConditionsParameter;
        Log.d("TimeConditions", "remaining time before alert: ", (timeConditionsParameter.getTimeToShowAlert() - SystemClock.elapsedRealtime()));
        return timeConditionsParameter.getTimeToShowAlert() < SystemClock.elapsedRealtime();
    }

    @Override
    public boolean isReadyForAction(InAppActionConditionsParameter inAppActionConditionsParameter) {
        return super.isReadyForAction(inAppActionConditionsParameter);
    }

    @Override
    public void updateInAppActionConditionsParameter(InAppActionConditionsParameter inAppActionConditionsParameter) {
        TimeConditionsParameter timeConditionsParameter = (TimeConditionsParameter) inAppActionConditionsParameter;
        long currentTime = SystemClock.elapsedRealtime();
        if (timeConditionsParameter.getLastDetectionTime() + 3000 < SystemClock.elapsedRealtime()) {
            timeConditionsParameter.setTimeToShowAlert(currentTime + timeBeforeCreatingAlert);
        }
        timeConditionsParameter.setLastDetectionTime(currentTime);
        super.updateInAppActionConditionsParameter(inAppActionConditionsParameter);
    }
}

File: ios/ObjectiveC/Beacon/3-Alert/Complete/beaconAlertComplete/TimeConditions.swift
//
//  TimeConditions.swift
//  beaconAlertComplete
//
//  Created by Connecthings on 14/11/2017.
//  Copyright © 2017 Connecthings. All rights reserved.
//

import Foundation
import ConnectPlaceActions
import ConnectPlaceCommon

public class TimeConditions: InAppActionConditionsDefault {
    @objc let delayBeforeCreation: Double
    
    public init(maxTimeBeforeReset: Double, delayBeforeCreation: Double) {
        self.delayBeforeCreation = delayBeforeCreation
        super.init(maxTimeBeforeReset: maxTimeBeforeReset)
    }
    
    public override func getApplicationNamespace() -> String {
        return String(reflecting: TimeConditions.self).split(separator: ".")[0].description
    }
    
    public override func getConditionsKey() -> String {
        return "inAppActionTimeConditions"
    }
    
    public override func getInAppActionParameterClass() -> String {
        return "TimeConditionsParameter"
    }
    
    public override func updateInAppActionConditionsParameter(_ conditionsParameter: InAppActionConditionsParameter) {
        if let strategyParameter = conditionsParameter as? TimeConditionsParameter {
            let currentTime: Double = CFAbsoluteTimeGetCurrent()
            if strategyParameter.lastDetectionTime + 3 < currentTime {
                GlobalLogger.shared.debug("delay before creating " , delayBeforeCreation)
                strategyParameter.timeToShowAlert = currentTime + delayBeforeCreation
            }
            strategyParameter.lastDetectionTime = currentTime + 0
        }
        super.updateInAppActionConditionsParameter(conditionsParameter)
    }
    
    public override func isConditionValid(_ conditionsParameter: InAppActionConditionsParameter) -> Bool {
        if let strategyParameter = conditionsParameter as? TimeConditionsParameter {
            GlobalLogger.shared.debug("delay remaining " , ( CFAbsoluteTimeGetCurrent() - strategyParameter.timeToShowAlert))
            return strategyParameter.timeToShowAlert < CFAbsoluteTimeGetCurrent()
        }
        return true
    }
}
